File: research/joinery/corpus/ledger.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

type Ledger struct {
	rows  []int64
	tags  map[string]int
	total *int64
}

func NewLedger() *Ledger {
	return &Ledger{tags: make(map[string]int)}
}

func (l *Ledger) Push(tag string, v int64) int {
	at := len(l.rows)
	l.rows = append(l.rows, v)
	l.tags[tag] = at
	l.total = nil
	return at
}

func (l *Ledger) Total() int64 {
	if l.total != nil {
		return *l.total
	}
	var acc int64
	for _, r := range l.rows {
		if r > 0 {
			acc += r
		}
	}
	l.total = &acc
	return acc
}

func main() {
	l := NewLedger()
	for i, arg := range os.Args[1:] {
		v, err := strconv.ParseInt(arg, 10, 64)
		if err != nil {
			fmt.Fprintf(os.Stderr, "skipping %s: %v\n", arg, err)
			continue
		}
		l.Push(fmt.Sprintf("arg%d", i), v)
	}
	fmt.Printf("total=%d\n", l.Total())
}

File: research/joinery/corpus/ledger.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct Ledger {
    rows: Vec<i64>,
    tags: HashMap<String, usize>,
    total: Option<i64>,
}

impl Ledger {
    pub fn new() -> Self {
        Ledger { rows: Vec::new(), tags: HashMap::new(), total: None }
    }

    pub fn push(&mut self, tag: &str, v: i64) -> usize {
        let at = self.rows.len();
        self.rows.push(v);
        self.tags.insert(tag.to_string(), at);
        self.total = None;
        at
    }

    pub fn total(&mut self) -> i64 {
        if let Some(t) = self.total {
            return t;
        }
        let t = self.rows.iter().filter(|&&r| r > 0).sum();
        self.total = Some(t);
        t
    }
}

fn main() {
    let mut l = Ledger::new();
    for (i, arg) in std::env::args().skip(1).enumerate() {
        match arg.parse::<i64>() {
            Ok(v) => { l.push(&format!("arg{i}"), v); }
            Err(e) => eprintln!("skipping {arg}: {e}"),
        }
    }
    println!("total={}", l.total());
}
